tests/test_model.py

```diff
@@ -99,6 +112,45 @@ def test_forward_infer(model, x):
     return model.apply(x)
 
 
-@specialize_std((make_model(), rand(3, 10)))
+@parse_compare((make_model(), rand(3, 10)), array=True)
 def test_forward_specialize(model, x):
     return model.apply(x)
+
+
+@infer_std(
+    type=[
+        ({'value': make_model()},
+         {'value': rand(3, 10)},
+         {'value': rand(3, 8)},
+         Model_t),
+        ({'value': make_model('float32')},
+         {'value': rand(3, 10)},
+         {'value': rand(3, 8)},
+         InferenceError),
+        ({'value': make_model('float32')},
+         {'value': rand(3, 10, dtype='float32')},
+         {'value': rand(3, 8, dtype='float32')},
+         Model_t_f32),
+    ],
+    shape=[
+        ({'value': make_model()},
+         {'value': rand(3, 10)},
+         {'value': rand(3, 8)},
+         ClassShape({
+             'layers': TupleShape((ClassShape({'W': (10, 15), 'b': (1, 15)}),
+                                   ClassShape({'W': (15, 8), 'b': (1, 8)})))})
+         ),
+        ({'value': make_model()},
+         {'value': rand(3, 14)},
+         {'value': rand(3, 8)},
+         InferenceError),
+    ]
+)
+def test_backward_infer(model, x, y):
+    return grad(cost)(model, x, y)
+
+
+@parse_compare((make_model(), rand(3, 10), rand(3, 8)),
+               array=True, python=False)
```

Comment:
That tests that the model runs, but it doesn't test that its output is sensible, does it? We should use `grad_test`, although it appears we can't use it directly, it'll have to be adapted.

Reply:
Yes, I agree with everything.